pkg/testutils/kubernetes/builders/client.go

```diff
@@ -0,0 +1,212 @@
+	WithContext(ctx context.Context) ClientBuilder
+	// WithErrorChannel sets a channel for reporting errors from observers
+	WithErrorChannel(chan error) ClientBuilder
+}
+
+// clientBuilder builds fake instances of Kubernetes.Interface
+type clientBuilder struct {
+	ctx       context.Context
+	client    *fake.Clientset
+	errors    []error
+	errCh     chan error
+	observers []func()
+	ready     sync.WaitGroup
+}
+
+// NewClientBuilder returns a ClientBuilder
+func NewClientBuilder() ClientBuilder {
+	return &clientBuilder{
+		ctx:    context.TODO(),
+		client: fake.NewSimpleClientset(),
+		errCh:  make(chan error),
+	}
+}
+
+func (b *clientBuilder) WithContext(ctx context.Context) ClientBuilder {
+	b.ctx = ctx
+	return b
+}
+
+func (b *clientBuilder) WithErrorChannel(errCh chan error) ClientBuilder {
+	b.errCh = errCh
+	return b
+}
+
+func (b *clientBuilder) WithObjects(objs ...runtime.Object) ClientBuilder {
+	for _, o := range objs {
+		err := b.client.Tracker().Add(o)
+		if err != nil {
+			b.errors = append(b.errors, err)
+			break
+		}
+	}
+	return b
+}
+
+func (b *clientBuilder) WithPods(pods ...*corev1.Pod) ClientBuilder {
+	for _, p := range pods {
+		_, err := b.client.CoreV1().Pods(p.Namespace).Create(context.TODO(), p, metav1.CreateOptions{})
+		if err != nil {
+			b.errors = append(b.errors, err)
+		}
+	}
+	return b
+}
+
+func (b *clientBuilder) WithNamespace(namespace string) ClientBuilder {
+	ns := &corev1.Namespace{
+		ObjectMeta: metav1.ObjectMeta{Name: namespace},
+	}
+	_, err := b.client.CoreV1().Namespaces().Create(context.TODO(), ns, metav1.CreateOptions{})
+	if err != nil {
+		b.errors = append(b.errors, err)
+	}
+
+	return b
+}
+
+func (b *clientBuilder) WithServices(services ...*corev1.Service) ClientBuilder {
+	for _, s := range services {
+		_, err := b.client.CoreV1().Services(s.Namespace).Create(context.TODO(), s, metav1.CreateOptions{})
+		if err != nil {
+			b.errors = append(b.errors, err)
+		}
+	}
+	return b
+}
+
+//nolint:gocognit,nestif
+func (b *clientBuilder) WithPodObserver(namespace string, event ObjectEvent, observer PodObserver) ClientBuilder {
+	// goroutine that watches for events that match the observer's subscription
+	obsFunc := func() {
+		watcher, err := b.client.CoreV1().Pods(namespace).Watch(
+			context.TODO(),
+			metav1.ListOptions{},
+		)
+		if err != nil {
+			b.errCh <- fmt.Errorf("failed watching pods : %w", err)
+			return
+		}
+		defer watcher.Stop()
+
+		b.ready.Done()
+		for {
+			select {
+			case <-b.ctx.Done():
+				return
+			case watcherEvent := <-watcher.ResultChan():
+				if watcherEvent.Type == watch.Error {
+					b.errCh <- fmt.Errorf("error event received watching pods")
+					return
+				}
+				pod, isPod := watcherEvent.Object.(*corev1.Pod)
+
+				// if we receive an unexpected object, ignore
+				if !isPod {
+					continue
+				}
+				objectEvent := ObjectEvent(watcherEvent.Type)
+				if event == ObjectEventAll || objectEvent == event {
```

Comment:
I'd suggest to invert this if and `continue`, ad oppose to increase indentation for the test of the function.

Reply:
I generally try to minimize indentation, but in this case, I found the condition becomes less readeable. 
```
if event != ObjetEventAll && objectEvent != event {
      continue
}
```

Reply:
🤔 I don't know, it does not seem that unreadable to me. I'd still prefer it.

Reply:
I understand that many indentation levels are not a good practice, but in this case, it is only one and If I have to trade between this condition with two negation and one indentation, I would prefer the indentations.  It seems to me a matter of preferences, don't you think?

Reply:
I believe most PR discussions are (and should be) a matter of preference, but that does not make them less valid. If we could coalesce the entirety of our intuition and experience in hard requirements, we could skip them altogether and just use linters :)

But I digress. Non-negated (pun intended) conditions are indeed more readable, and although I think returning early is preferred, I'm not adamant on it.

Reply:
> I believe most PR discussions are (and should be) a matter of preference

And I'm convinced they should not be about preferences. PR reviews should not focus not in making the code more to the reviewer's taste but to ensure we are not infringing damage to our future selves.

Reply:
I just realized I explained very poorly my point in my previous comments when I said it was about preferences on indentation over the alternative if condition.

What I meant to say is that in this context, my criteria was that double negation has a larger cognitive load than a single indentation. So, in reality, I wasn't talking about preferences but about perceptions, that are still subjective but not the same.